File: app/src/main/java/com/github/wyang/basiclibrarydemo/ComplexViewActivity.java
package com.github.wyang.basiclibrarydemo;

import android.support.v7.widget.SwitchCompat;
import android.support.v7.widget.Toolbar;
import android.util.Log;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.github.android.common.widget.shape.extend.ComplexView;

import butterknife.BindView;
import butterknife.OnClick;

/**
 * Created by weiyang on 2019-10-12.
 *
 * @url https://www.jb51.net/article/145309.htm
 */
public class ComplexViewActivity extends BaseActivity {
    @BindView(R.id.toolbar)
    Toolbar toolbar;

    @BindView(R.id.sc_enable)
    SwitchCompat sc_enable;

    @BindView(R.id.ll_group)
    LinearLayout ll_group;

    @Override
    public int getLayoutId() {
        return R.layout.activity_complex_view;
    }

    @Override
    protected void initView() {
        initToolBar(toolbar, "ComplexView", true);

        sc_enable.setOnCheckedChangeListener((buttonView, isChecked) -> {
            ViewGroup parent = (ViewGroup) sc_enable.getParent();
            applyEnable(parent, isChecked);
        });

        setSelectListener(ll_group);
    }

    private void applyEnable(View view, boolean enable) {
        if (view instanceof ViewGroup) {
            ViewGroup parent = (ViewGroup) view;
            for (int i = 0; i < parent.getChildCount(); i++) {
                applyEnable(parent.getChildAt(i), enable);
            }
        } else {
            if (!(view instanceof SwitchCompat))
                view.setEnabled(enable);
        }
    }

    private void setSelectListener(View view) {
        if (view instanceof LinearLayout) {
            LinearLayout group = (LinearLayout) view;
            for (int i = 0; i < group.getChildCount(); i++) {
                setSelectListener(group.getChildAt(i));
            }
        } else if (view instanceof ComplexView) {
            view.setOnClickListener(v -> {
                applySelected(ll_group, false);
                v.setSelected(true);
            });
        }
    }

    private void applySelected(View view, boolean selected) {
        if (view instanceof LinearLayout) {
            LinearLayout parent = (LinearLayout) view;
            for (int i = 0; i < parent.getChildCount(); i++) {
                applySelected(parent.getChildAt(i), selected);
            }
        } else if (view instanceof ComplexView) {
            view.setSelected(selected);
        }
    }

    @BindView(R.id.cv_prompt)
    ComplexView cv_prompt;

    @OnClick(R.id.cv_prompt)
    public void onClickPrompt() {
        cv_prompt.setSelected(!cv_prompt.isSelected());
        cv_prompt.setText(cv_prompt.isSelected() ? "已启用" : "启用提醒");
    }
}

File: app/src/main/java/com/github/wyang/basiclibrarydemo/ComplexViewActivity.kt
package com.github.wyang.basiclibrarydemo

import android.view.View
import android.view.ViewGroup
import android.widget.CompoundButton
import android.widget.LinearLayout
import androidx.appcompat.widget.SwitchCompat
import com.github.android.common.widget.shape.extend.ComplexView
import com.github.wyang.basiclibrarydemo.databinding.ActivityComplexViewBinding
import com.github.wyang.basiclibrarydemo.databinding.LayoutDemoButton1Binding
import com.github.wyang.basiclibrarydemo.databinding.LayoutDemoButton2Binding
import com.github.wyang.basiclibrarydemo.databinding.LayoutToolbarBinding

class ComplexViewActivity : BaseActivity<ActivityComplexViewBinding>() {
    private var toolbarBinding: LayoutToolbarBinding? = null
    private var layout1Binding: LayoutDemoButton1Binding? = null
    private var layout2Binding: LayoutDemoButton2Binding? = null

    override fun getViewBinding(): ActivityComplexViewBinding {
        val binding = ActivityComplexViewBinding.inflate(layoutInflater)
        toolbarBinding = binding.includeToolbar
        layout1Binding = binding.includeLayout1
        layout2Binding = binding.includeLayout2
        return binding
    }

    override fun initialize() {
        toolbarBinding?.let { initToolBar(it.toolbar, "ComplexView", true) };

        layout1Binding?.scEnable?.setOnCheckedChangeListener { _: CompoundButton?, isChecked: Boolean ->
            val parent = layout1Binding?.scEnable?.parent as ViewGroup
            applyEnable(parent, isChecked)
        }

        layout2Binding?.let { setSelectListener(it.llGroup) }


        layout2Binding?.cvPrompt?.setOnClickListener {
            layout2Binding?.cvPrompt?.isSelected = !layout2Binding?.cvPrompt?.isSelected!!
            layout2Binding?.cvPrompt?.text = if (layout2Binding?.cvPrompt?.isSelected!!) "已启用" else "启用提醒"
        }
    }

    private fun applyEnable(view: View, enable: Boolean) {
        if (view is ViewGroup) {
            for (i in 0 until view.childCount) {
                applyEnable(view.getChildAt(i), enable)
            }
        } else {
            if (view !is SwitchCompat) view.isEnabled = enable
        }
    }

    private fun setSelectListener(view: View) {
        if (view is LinearLayout) {
            for (i in 0 until view.childCount) {
                setSelectListener(view.getChildAt(i))
            }
        } else (view as? ComplexView)?.setOnClickListener { v: View ->
            layout2Binding?.let { applySelected(it.llGroup, false) }
            v.isSelected = true
        }
    }

    private fun applySelected(view: View, selected: Boolean) {
        if (view is LinearLayout) {
            for (i in 0 until view.childCount) {
                applySelected(view.getChildAt(i), selected)
            }
        } else (view as? ComplexView)?.isSelected = selected
    }

}